File: tests/test_bundle.js
#!/usr/bin/env node
/**
 * Node.js test script for the CityDatabase module
 */

const fs = require('fs');
const path = require('path');

if (process.argv.length < 3) {
    console.error('Usage: node test.js <path_to_city_database>');
    process.exit(1);
}

// Import the CityDatabase class from CLI args
const dbPath = process.argv[2]
const CityDatabase = require('../cityDatabase');

async function runTests() {
    console.log('Testing CityDatabase module in Node.js...\n');

    // Load the bundle file
    if (!fs.existsSync(dbPath)) {
        console.error(`Bundle file not found: ${dbPath}`);
        console.error('Run "python scripts/bundle.py" first to create the bundle.');
        process.exit(1);
    }

    console.log('Loading database bundle...');
    const buffer = fs.readFileSync(dbPath);

    const db = new CityDatabase();
    await db.loadFromBundle(buffer.buffer);

    console.log('Database loaded successfully!\n');

    // Show statistics
    const stats = db.getStats();
    console.log('Database Statistics:');
    console.log(`  Cities: ${stats.cities.toLocaleString()}`);
    console.log(`  Regions: ${stats.regions.toLocaleString()}`);
    console.log(`  Countries: ${stats.countries.toLocaleString()}`);
    console.log(`  Vocabulary: ${stats.vocabularySize.toLocaleString()} tokens\n`);

    // Test specific geohashes
    const testCases = [
        'u14zy',    // 's-Gravenzande
        'u15y0',    // 's-Hertogenbosch  
        's6ruxt',   // 'Ārdamatā
        'stq1d',    // Another city
        'ezdn8',    // Another city
        'nonexistent' // Should return null
    ];

    console.log('Testing specific geohashes:');
    for (const geohash of testCases) {
        const result = db.getCityFromGeohash(geohash);
        if (result) {
            console.log(`  ${geohash}: ${result}`);
        } else {
            console.log(`  ${geohash}: Not found`);
        }
    }

    console.log('\nTesting prefix search for "u1":');
    const prefixResults = db.findCitiesByGeohashPrefix('u1', 5);
    for (const city of prefixResults) {
        const formatted = db.getCityFromGeohash(city.geohash);
        console.log(`  ${city.geohash}: ${formatted}`);
    }

    // Performance test
    console.log('\nPerformance test (1000 random lookups):');
    const startTime = Date.now();
    let found = 0;

    for (let i = 0; i < 1000; i++) {
        // Generate a random geohash-like string
        const randomHash = Math.random().toString(36).substring(2, 8);
        const result = db.getCityFromGeohash(randomHash);
        if (result) found++;
    }

    const endTime = Date.now();
    const duration = endTime - startTime;
    console.log(`  Time: ${duration}ms for 1000 lookups`);
    console.log(`  Average: ${(duration / 1000).toFixed(2)}ms per lookup`);
    console.log(`  Found: ${found} cities out of 1000 random hashes`);

    // Test with actual geohashes for performance
    console.log('\nPerformance test with real geohashes:');
    const realHashes = db.cities.slice(0, 1000).map(city => city.geohash);
    const startTime2 = Date.now();

    for (const hash of realHashes) {
        db.getCityFromGeohash(hash);
    }

    const endTime2 = Date.now();
    const duration2 = endTime2 - startTime2;
    console.log(`  Time: ${duration2}ms for 1000 real geohash lookups`);
    console.log(`  Average: ${(duration2 / 1000).toFixed(3)}ms per lookup`);

    console.log('\n✅ All tests completed successfully!');
}

// Run the tests
runTests().catch(error => {
    console.error('❌ Test failed:', error);
    process.exit(1);
});

File: tests/test_bundle.py
#!/usr/bin/env python3
"""
Verify the bundle format and test basic decompression in Python
"""

import struct
import os
import sys


def verify_bundle(bundle_path):
    """Verify the bundle file structure"""
    print(f"Verifying bundle: {bundle_path}")

    if not os.path.exists(bundle_path):
        print(f"ERROR: Bundle file not found: {bundle_path}")
        return False

    file_size = os.path.getsize(bundle_path)
    print(f"Bundle size: {file_size:,} bytes ({file_size / 1024 / 1024:.2f} MB)")

    with open(bundle_path, "rb") as f:
        # Read header
        magic = f.read(8).decode("ascii")
        if magic != "CITYDB01":
            print(f"ERROR: Invalid magic bytes: {magic}")
            return False
        print(f"✓ Magic bytes: {magic}")

        num_vocab = struct.unpack("<I", f.read(4))[0]
        num_cities = struct.unpack("<I", f.read(4))[0]
        num_regions = struct.unpack("<I", f.read(4))[0]
        num_countries = struct.unpack("<I", f.read(4))[0]
        reserved = f.read(8)

        print(f"✓ Vocabulary entries: {num_vocab:,}")
        print(f"✓ Cities: {num_cities:,}")
        print(f"✓ Regions: {num_regions:,}")
        print(f"✓ Countries: {num_countries:,}")

        # Read vocabulary
        print("Reading vocabulary...")
        vocab = {}
        for i in range(num_vocab):
            token_id = struct.unpack(">H", f.read(2))[0]
            token_len = struct.unpack("B", f.read(1))[0]
            token_text = f.read(token_len).decode("utf-8")
            vocab[token_id] = token_text

            if i < 10:  # Show first 10 tokens
                print(f"  {token_id}: '{token_text}'")
            elif i == 10:
                print("  ...")

        print(f"✓ Loaded {len(vocab)} vocabulary entries")

        # Read data sections
        sections = ["geohash", "regionIds", "countryIds", "cityTokens", "regionTokens", "countryTokens"]

        print("Reading data sections...")
        for section in sections:
            section_size = struct.unpack("<I", f.read(4))[0]
            section_data = f.read(section_size)
            print(f"  {section}: {section_size:,} bytes")

        remaining = file_size - f.tell()
        if remaining != 0:
            print(f"WARNING: {remaining} bytes remaining in file")
        else:
            print("✓ Entire file consumed")

    return True


def test_basic_decompression(bundle_path):
    """Test basic decompression of a few entries"""
    print("\nTesting basic decompression...")

    return verify_bundle(bundle_path)


def main():
    print("Bundle verification and basic decompression test\n")

    if len(sys.argv) > 1:
        bundle_path = sys.argv[1]
    else:
        print("Error: Please provide the path to the bundle file as an argument")
        return 1

    if test_basic_decompression(bundle_path):
        print("\n✅ Bundle verification passed!")
        print("\nTo test the JavaScript module:")
        print("1. Open http://localhost:8000/test.html in your browser")
        print("2. Click 'Load Database' to load the bundle")
        print("3. Test the geohash lookup functions")
        print("\nExample test geohashes:")
        print("  u14zy - Should find 's-Gravenzande")
        print("  u15y0 - Should find 's-Hertogenbosch")
        print("  s6ruxt - Should find 'Ārdamatā")
    else:
        print("\n❌ Bundle verification failed!")
        return 1

    return 0


if __name__ == "__main__":
    exit(main())
